### ISSUE TITLE
SIGSEGV on MIPS when dividing two 128-bit integers

### ISSUE THREAD

**Issue by japaric** [rust-lang/rust#41222]
### STR

``` rust
#![feature(i128_type)]

fn main() {
    let x: u128 = 1844674407370955161400;
    let y: u128 = 18002795302052819272;
    let z = x / y;

    println!("{}", x);
    println!("{}", y);
    println!("{}", z);
}
```

```
$ cargo run
1844674407370955161400
18002795302052819272
102

$ cross run --target mips-unknown-linux-gnu
     Running `/target/mips-unknown-linux-gnu/debug/hello`
Segmentation fault (core dumped)
```

This operation / intrinsic is implemented in Rust. Other 32-bit targets don't have this problem.

```
$ cross run --target arm-unknown-linux-gnueabi
1844674407370955161400
18002795302052819272
102
```

### Meta

```
$ rustc -V
rustc 1.18.0-nightly (3b5754e5c 2017-04-10)
```

cc @est31 I suspect this is some ABI problem.

**Comment by nagisa** [comment]
cc me

**Comment by nagisa** [comment]
The result of the division is correct. Invalid memory access happens during printing of the value.

**Comment by nagisa** [comment]
Okay it might not be correct in that the 2 halves may be swapped. Printing sigsegvs for code as simple as

Println!("{}", 102u128<<64)

**Comment by nagisa** [comment]
Division is confirmed to be correct. There's two issues though. Parsing and printing i128 is broken for values over 64 bits.

**Comment by nikomatsakis** [comment]
triage: P-high

Assigning to @nagisa.

**Comment by nagisa** [comment]
I have tried the new nightly which i think also includes the division fix from the other PR and it still sigsegvs. Will investigate more.

**Comment by nagisa** [comment]
```
#![feature(i128_type)]

fn x(i: u128) -> isize {
    (i % 10000) as isize
}

fn main() {
    let y = (0x1u128 << 64);
    println!("{}", x(y));
}
```

prints a `-10000`, which is obviously wrong (the expected result would be `1616`).

```
#![feature(i128_type)]
fn x(i: u128) -> u128 {
    (i % 10000)

}

fn main() {
    let y = (0x1u128 << 64);
    println!("{:x} {:x}", (x(y) >> 64) as u64, x(y) as u64);
}
```

prints `1 ffffffffffffd8f0`.

ABI is fine (identity works), so the blame falls squarely onto implementation of the intrinsics. i686 works fine as well, so it is not a 32-bit vs 64-bit problem.

**Comment by nagisa** [comment]
This is a bug in LLVM generated code when handling overflowing addition/subtraction operations. `add i128 0, -1` results in `0x0000_0000_0000_0000_ffff_ffff_ffff_ffff`. I will report and/or investigate fixing it later on.

**Comment by nagisa** [comment]
Filled https://bugs.llvm.org/show_bug.cgi?id=32713

**Comment by nikomatsakis** [comment]
triage: P-medium

We're blocked on LLVM here, and MIPS is not a tier 1 platform or anything like that.

**Comment by nagisa** [comment]
https://reviews.llvm.org/rL305389

**Comment by nagisa** [comment]
Do we want to consider backporting this? I don’t think this is critical enough to bother.

**Comment by arielb1** [comment]
@nagisa 

I think this is important enough to backport into nightly at the same time I backport the ARMv6 fixes.

**Comment by est31** [comment]
With the update to LLVM 6.0 is this issue resolved now? @japaric can you check?

**Comment by nagisa** [comment]
This has been solved with fixes being present since upstream LLVM 5.0, to my knowledge.
